servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/KeepAliveManager.java

```diff
@@ -0,0 +1,321 @@
+                     final Scheduler scheduler, final IdlenessDetector idlenessDetector) {
+        this.channel = channel;
+        this.scheduler = scheduler;
+        if (keepAlivePolicy != null) {
+            disallowKeepAliveWithoutActiveStreams = !keepAlivePolicy.withoutActiveStreams();
+            pingAckTimeoutMillis = keepAlivePolicy.ackTimeout().toMillis();
+            final GenericFutureListener<Future<? super Void>> goAwayListener = f -> {
+                if (f.isSuccess()) {
+                    LOGGER.debug("Closing channel={}, after keep-alive timeout.", this.channel);
+                    KeepAliveManager.this.close0();
+                }
+            };
+            pingWriteCompletionListener = future -> {
+                if (future.isSuccess() && keepAliveState == KEEP_ALIVE_ACK_PENDING) {
+                    // Schedule a task to verify ping ack within the pingAckTimeoutMillis
+                    keepAliveState = scheduler.afterMillis(() -> {
+                        if (keepAliveState != null) {
+                            keepAliveState = KEEP_ALIVE_ACK_TIMEDOUT;
+                            LOGGER.debug(
+                                    "channel={}, timeout {}ms waiting for keep-alive PING(ACK), writing go_away.",
+                                    this.channel, pingAckTimeoutMillis);
+                            channel.writeAndFlush(new DefaultHttp2GoAwayFrame(NO_ERROR))
+                                    .addListener(goAwayListener);
+                        }
+                    }, pingAckTimeoutMillis);
+                }
+            };
+            int idleInSeconds = (int) keepAlivePolicy.idleDuration().getSeconds();
+            idlenessDetector.configure(channel, idleInSeconds, this::channelIdle);
+        } else {
+            disallowKeepAliveWithoutActiveStreams = false;
+            pingAckTimeoutMillis = DEFAULT_ACK_TIMEOUT.toMillis();
+            pingWriteCompletionListener = null;
+        }
+    }
+
+    void pingReceived(final Http2PingFrame pingFrame) {
+        assert channel.eventLoop().inEventLoop();
+
+        if (pingFrame.ack()) {
+            long pingAckContent = pingFrame.content();
+            if (pingAckContent == GRACEFUL_CLOSE_PING_CONTENT) {
+                LOGGER.debug("channel={}, graceful close ping ack received.", channel);
+                cancelIfStateIsAFuture(gracefulCloseState);
+                gracefulCloseWriteSecondGoAway();
+            } else if (pingAckContent == KEEP_ALIVE_PING_CONTENT) {
+                cancelIfStateIsAFuture(keepAliveState);
+                keepAliveState = null;
+            }
+        } else {
+            // Send an ack for the received ping
+            channel.writeAndFlush(new DefaultHttp2PingFrame(pingFrame.content(), true));
+        }
+    }
+
+    void trackActiveStream(final Channel streamChannel) {
+        activeChildChannelsUpdater.incrementAndGet(this);
```

Comment:
If we change this code to always use `addListener` (doesn't look it's practical in real network application that promise will complete so fast):
https://github.com/apple/servicetalk/blob/b22e5ea082e3fba2bd0e97d7ef416941301f835f/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/H2ClientParentConnectionContext.java#L229-L234
Then this method will always be on the event loop. `closeFuture` also invokes listeners on the event loop. We will be able to make this variable non-volatile then. WDYT? Can be a follow-up as it's not related to this PR.

Reply:
I am not sure about the API inter-connects here and would like to defer.